network/wsPeer.go

```diff
@@ -571,15 +581,34 @@ func (wp *wsPeer) readLoop() {
 			networkP2PReceivedBytesByTag.Add(string(tag[:]), uint64(len(msg.Data)+2))
 			networkP2PMessageReceivedByTag.Add(string(tag[:]), 1)
 		}
-		msg.Data, err = dataConverter.convert(msg.Tag, msg.Data)
+		originalTag := msg.Tag
+		msg.Data, err = wp.msgCodec.decompress(msg.Tag, msg.Data)
 		if err != nil {
+			// Handle VP errors by sending abort message and continuing
+			var vcErr *voteCompressionError
+			if errors.As(err, &vcErr) {
+				if close, reason := wp.handleVPError(err); close {
+					cleanupCloseError = reason
+					return
+				}
+				// Drop this vote and continue reading
+				continue
+			}
+			// Non-VP errors tear down connection
 			wp.reportReadErr(err)
 			return
 		}
+		// If decompress returned nil (e.g., for abort message), drop the message
+		if msg.Data == nil {
+			continue
+		}
 		if wp.peerType == peerTypeWs {
-			networkReceivedUncompressedBytesByTag.Add(string(tag[:]), uint64(len(msg.Data)+2))
+			networkReceivedUncompressedBytesByTag.Add(string(msg.Tag), uint64(len(msg.Data)+2))
 		} else {
-			networkP2PReceivedUncompressedBytesByTag.Add(string(tag[:]), uint64(len(msg.Data)+2))
+			networkP2PReceivedUncompressedBytesByTag.Add(string(msg.Tag), uint64(len(msg.Data)+2))
+		}
+		if originalTag == protocol.VotePackedTag {
```

Comment:
Why do you need `originalTag` to be saved way earlier (or at all?)
Since the only place you need it is to check for the change,
```
if msg.Tag == protocol.VotePackedTag { 
```
works, doesn't it?